### PR TITLE
Macho dos (dont squash this PR)

### PR DESCRIPTION
**Pull request by trufae**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

The binary in the testbins repo makes r2 take forever to load and eventually run out of memory or crash. fixing this bug results in a null deref when deintializing the rbinobject. this pr fixes both bugs.

The 3rd commit fixes a null deref when loading machos without segments

**Test plan**

See https://github.com/radareorg/radare2-testbins/pull/28

**Closing issues**

One from clusterfuzz

### CODE REVIEW

File: libr/bin/format/mach0/mach0.c [hunk 1]
```diff
@@ -3234,6 +3236,9 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
 		size_t partition_sizes[] = {bind_size, lazy_size, weak_size};
 		size_t pidx;
 		int opcodes_offset = 0;
+		if (!bin->segs) {
+			return NULL;
```
Comment (mrmacete) [inline libr/bin/format/mach0/mach0.c:3240]: looks like it's leaking `relocs` and `opcodes`

File: libr/bin/format/mach0/mach0.c [hunk 2]
```diff
@@ -4267,7 +4272,7 @@ RList *MACH0_(mach_fields)(RBinFile *bf) {
 			ut32 nsects = r_buf_read_le32_at (buf, addr + (is64 ? 64 : 48));
 			ut64 off = is64 ? 72 : 56;
 			int j = 0;
-			while (off < lcSize && nsects--) {
+			while (off < lcSize && (addr + off) < length && nsects-- > 0) {
```
Comment (wargio) [inline libr/bin/format/mach0/mach0.c:4275]: please use a for if you need to decrement. makes the code more readable and prevents weird usecases.

**Comment by mrmacete** [comment]
oops apparently there's a breaking change in the binfield refactoring

File: libr/bin/format/mach0/mach0.c [hunk 3]
```diff
@@ -4266,15 +4271,12 @@ RList *MACH0_(mach_fields)(RBinFile *bf) {
...
-				if (is64) {
-					r_list_append (ret, r_bin_field_new (addr + off, addr + off, 1, sdb_fmt ("section_%d", j++), "mach0_section64", "mach0_section64", true));
-					off += 80;
-				} else {
-					r_list_append (ret, r_bin_field_new (addr + off, addr + off, 1, sdb_fmt ("section_%d", j++), "mach0_section", "mach0_section", true));
-					off += 68;
-				}
+			size_t i, j = 0;
+			for (i = 0; i < nsects && (addr + off) < length && off < lcSize; i++) {
+				const char *section_name = is64? "mach0_section64": "mach0_section";
```
Comment (mrmacete) [inline libr/bin/format/mach0/mach0.c:4276]: not used?

**Pull request merged**